rbac/migration_tool/migrate.py

```diff
@@ -108,46 +49,44 @@ def migrate_groups_for_tenant(tenant: Tenant, replicator: RelationReplicator):
+            # and the group must be locked to add principals to the groups.
+            # NOTE: The lock on the group is not necessary when adding system roles to the group,
+            # as the binding mappings are locked during this process to ensure concurrency control.
+            # Start of transaction for group operations
             with transaction.atomic():
+                # Lock group
+                Group.objects.select_for_update().get(pk=group.pk)
                 dual_write_handler = RelationApiDualWriteGroupHandler(
                     group, ReplicationEventType.MIGRATE_TENANT_GROUPS, replicator=replicator
                 )
+                # this operation requires lock on group as well as in view,
+                # more details in GroupViewSet#get_queryset method which is used to add principals.
                 dual_write_handler.generate_relations_to_add_principals(principals)
+                # lock on group is not required to add system role, only binding mappings which is included in
+                # dual_write_handler
                 dual_write_handler.generate_relations_to_add_roles(system_roles)
                 dual_write_handler.replicate()
+            # End of transaction for group operations, locks are released
 
 
 def migrate_roles_for_tenant(tenant, exclude_apps, replicator):
     """Migrate all roles for a given tenant."""
-    default_workspace = Workspace.objects.get(type=Workspace.Types.DEFAULT, tenant=tenant)
-
     roles = tenant.role_set.all()
     if exclude_apps:
         roles = roles.exclude(access__permission__application__in=exclude_apps)
 
     for role in roles:
         logger.info(f"Migrating role: {role.name} with UUID {role.uuid}.")
 
-        tuples, mappings = migrate_role(role, default_workspace)
-
-        # Conflicts are not ignored in order to prevent this from
-        # accidentally running concurrently with dual-writes.
-        # If migration should be rerun, then the bindings table should be dropped.
-        # If changing this to allow updates,
-        # always ensure writes are paused before running.
-        # This must always be the case, but this should at least start failing you if you forget.
-        BindingMapping.objects.bulk_create(mappings, ignore_conflicts=False)
-
-        replicator.replicate(
-            ReplicationEvent(
-                event_type=ReplicationEventType.MIGRATE_CUSTOM_ROLE,
-                info={"role_uuid": str(role.uuid)},
-                partition_key=PartitionKey.byEnvironment(),
-                add=tuples,
+        with transaction.atomic():
+            role = Role.objects.select_for_update().get(pk=role.pk)
+            dual_write_handler = RelationApiDualWriteHandler(
+                role, ReplicationEventType.MIGRATE_CUSTOM_ROLE, replicator
             )
-        )
+
+            dual_write_handler.replicate_new_or_updated_role(role)
```

Comment:
Here I think we might need two adjustments:

- We need `dual_write_handler.prepare_for_update()` before the replicate method
- We might have a similar problem to above group principal removal with role deletion here. Normally, dual write shouldn't need to lock the role on deletion, because foreign key constraints on modifications to the role will prevent inconsistency. But here is another case where the migrator is not adding the data to the DB at the same time. So I think we could fix this one by _either_ re-running `.save()` on a role serializer here, so that access FKs serialize concurrent removal, OR we have to lock the role in the role view queryset on role deletion also.

Reply:
- I added `prepare_for_update`(I though that removal of current relations is not needed for role )
- I think lock on role  in delete action is already here